### PR TITLE
Mix precision training enabling

### PR DESCRIPTION
**Pull request by kprokofi**

### CODE REVIEW

File: torchreid/optim/sam.py [hunk 1]
```diff
@@ -69,3 +75,31 @@ def _grad_norm(self):
                     p=2
                )
         return norm
+
+    def has_overflow(self, params):
```
Comment (sovrasov) [inline torchreid/optim/sam.py]: Does this method belong to the public api of this class? If not, it could be marked with underscore.
Also two methods below could be static, since they actually don't use class members.

File: torchreid/engine/image/am_softmax.py [hunk 2]
```diff
@@ -267,22 +271,34 @@ def forward_backward(self, data):
 
             total_loss.backward(retain_graph=self.enable_metric_losses)
 
+            # backward pass
+            self.scaler.scale(total_loss).backward(retain_graph=False)
             for model_name in model_names:
-                if not self.models[model_name].training:
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py:271]: Why this was deleted? NNCF-related aux model freezing could be broken.

File: torchreid/engine/image/am_softmax.py [hunk 3]
```diff
@@ -267,22 +271,34 @@ def forward_backward(self, data):
...
-                    continue
-                if self.clip_grad != 0:
+                if self.clip_grad != 0 and step == 1:
+                    self.scaler.unscale_(self.optims[model_name])
                     torch.nn.utils.clip_grad_norm_(self.models[model_name].parameters(), self.clip_grad)
-                for trg_id in range(self.num_targets):
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py:275]: Metric losses are silently broken now

**Comment by kprokofi** [comment]
@ sovrasov, done

**Review by sovrasov — APPROVED** [review]
